File: ios/SylKit/Sources/SylKit/Model/Health.swift
import Foundation

/// Liveness and dependency status. The one endpoint that takes no bearer token.
public struct HealthStatus: Codable, Equatable, Sendable {
    public enum Status: String, Codable, Equatable, Sendable, CaseIterable {
        case ok, degraded, down
    }

    public let status: Status
    public let version: String
    public let startedAt: Date
    public let now: Date
    public let checks: [HealthCheck]
    /// What the answering process was BUILT FROM.
    ///
    /// `nil` when the service is running from source rather than from a build,
    /// which is a real answer and not a missing one. Read leniently and always
    /// written back with an explicit null, matching what the service emits —
    /// an absent field is indistinguishable from an older service that never
    /// had one, and "which build is answering" is exactly the question that
    /// must not be ambiguous.
    public let build: BuildInfo?
    /// How many turns are running or queued. Absent from a service with no
    /// conversation surface, so it is read and written as genuinely optional.
    public let turnsInFlight: Int?

    public init(
        status: Status,
        version: String,
        startedAt: Date,
        now: Date,
        checks: [HealthCheck],
        build: BuildInfo? = nil,
        turnsInFlight: Int? = nil
    ) {
        self.status = status
        self.version = version
        self.startedAt = startedAt
        self.now = now
        self.checks = checks
        self.build = build
        self.turnsInFlight = turnsInFlight
    }

    private enum CodingKeys: String, CodingKey {
        case status, version, startedAt, now, checks, build, turnsInFlight
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        status = try container.decode(Status.self, forKey: .status)
        version = try container.decode(String.self, forKey: .version)
        startedAt = try container.decode(Date.self, forKey: .startedAt)
        now = try container.decode(Date.self, forKey: .now)
        checks = try container.decode([HealthCheck].self, forKey: .checks)
        build = try container.decodeIfPresent(BuildInfo.self, forKey: .build)
        turnsInFlight = try container.decodeIfPresent(Int.self, forKey: .turnsInFlight)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(status, forKey: .status)
        try container.encode(version, forKey: .version)
        try container.encode(startedAt, forKey: .startedAt)
        try container.encode(now, forKey: .now)
        try container.encode(checks, forKey: .checks)
        try container.encodeRequiredNullable(build, forKey: .build)
        try container.encodeIfPresent(turnsInFlight, forKey: .turnsInFlight)
    }
}

/// Provenance for the artifact that is answering, stamped at build time.
///
/// The reason this exists at all: a stale build is invisible by construction.
/// Every health check passes, because the old build is perfectly healthy. It
/// cost three hours once — the service ran from 19:58 while a fix landed at
/// 20:18 — and it was noticed only because the Commander thought something read
/// oddly and asked.
public struct BuildInfo: Codable, Equatable, Sendable {
    /// The full SHA the build was made from, or `nil` when it was made outside
    /// a checkout — worth reporting rather than hiding.
    public let commit: String?
    public let builtAt: Date
    /// Whether the tree had uncommitted changes at build time. A dirty build
    /// cannot be reproduced from `commit`.
    public let dirty: Bool
    public let branch: String?

    public init(commit: String?, builtAt: Date, dirty: Bool, branch: String?) {
        self.commit = commit
        self.builtAt = builtAt
        self.dirty = dirty
        self.branch = branch
    }

    private enum CodingKeys: String, CodingKey {
        case commit, builtAt, dirty, branch
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        commit = try container.decodeIfPresent(String.self, forKey: .commit)
        builtAt = try container.decode(Date.self, forKey: .builtAt)
        dirty = try container.decode(Bool.self, forKey: .dirty)
        branch = try container.decodeIfPresent(String.self, forKey: .branch)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeRequiredNullable(commit, forKey: .commit)
        try container.encode(builtAt, forKey: .builtAt)
        try container.encode(dirty, forKey: .dirty)
        try container.encodeRequiredNullable(branch, forKey: .branch)
    }
}

public struct HealthCheck: Codable, Equatable, Sendable {
    public let name: String
    public let status: HealthStatus.Status
    /// Optional in the contract, not required-and-nullable — so it is read
    /// leniently and always written back with an explicit null, which is the
    /// shape the service actually emits.
    public let detail: String?

    public init(name: String, status: HealthStatus.Status, detail: String?) {
        self.name = name
        self.status = status
        self.detail = detail
    }

    private enum CodingKeys: String, CodingKey {
        case name, status, detail
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        name = try container.decode(String.self, forKey: .name)
        status = try container.decode(HealthStatus.Status.self, forKey: .status)
        detail = try container.decodeIfPresent(String.self, forKey: .detail)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(name, forKey: .name)
        try container.encode(status, forKey: .status)
        try container.encodeRequiredNullable(detail, forKey: .detail)
    }
}

File: backend/src/routes/health.ts
import { Router } from "express";

import type { BuildInfo, HealthCheck, HealthStatus } from "@syl/shared";

import type { SylConfig } from "../config.js";
import { instant, systemClock, type Clock } from "../services/clock.js";
import type { Database } from "../services/sqlite.js";
import { sendOk } from "./envelope.js";

/**
 * Liveness and dependency status, in the contract's shape.
 *
 * `status` is the worst of the checks rather than a separate judgement, so a
 * green top line can never sit above a red dependency. `degraded` is a real
 * and useful answer: Syl with a failing APNs channel is still worth talking
 * to, and reporting `down` would send a monitor into a restart loop over
 * something a restart cannot fix.
 *
 * The check that is genuinely ours is `subscription-rails`. Syl's strongest
 * constraint is that she runs on subscription payment rails and never the
 * metered API, and a stray `ANTHROPIC_API_KEY` in the environment silently
 * outranks the claude.ai login. A health endpoint that could only say "ok"
 * would report perfect health while the Commander's billing was rerouted —
 * this is how that hazard becomes visible before a statement makes it visible.
 */

/** One dependency this service can be asked about. */
export interface HealthProbe {
  readonly name: string;
  /**
   * Answer for this dependency. Must not throw and must not block — a health
   * endpoint that hangs is worse than one that reports `down`, because a
   * monitor cannot tell it apart from a network failure.
   */
  run(): { readonly status: HealthCheck["status"]; readonly detail?: string | null };
}

const SEVERITY: Readonly<Record<HealthCheck["status"], number>> = {
  ok: 0,
  degraded: 1,
  down: 2,
};

/** The worst status among the checks, or `ok` when there are none. */
export function worstStatus(
  checks: readonly HealthCheck[],
): HealthCheck["status"] {
  let worst: HealthCheck["status"] = "ok";
  for (const check of checks) {
    if (SEVERITY[check.status] > SEVERITY[worst]) worst = check.status;
  }
  return worst;
}

/**
 * The billing check, built from configuration alone.
 *
 * `degraded` rather than `down`: the service works, and the harness strips
 * both credential variables before spawning `claude`. What is at risk is
 * anything that reaches the CLI another way.
 */
export function subscriptionRailsProbe(config: SylConfig): HealthProbe {
  return {
    name: "subscription-rails",
    run: () =>
      config.subscriptionRails
        ? { status: "ok", detail: "claude.ai subscription" }
        : {
            status: "degraded",
            detail:
              `${config.credentialSource} is set. The harness strips it before spawning claude, ` +
              `but anything reaching the CLI another way bills the metered API.`,
          },
  };
}

/**
 * The store, checked by actually querying it.
 *
 * A probe that only asks whether a handle exists reports `ok` against a
 * database whose file has been deleted out from under it — SQLite keeps
 * serving from the open descriptor until it needs a page it does not have.
 */
export function databaseProbe(db: Database): HealthProbe {
  return {
    name: "database",
    run: () => {
      try {
        db.prepare("SELECT count(*) AS n FROM schema_migrations").get();
        return { status: "ok", detail: null };
      } catch (error) {
        return { status: "down", detail: error instanceof Error ? error.message : "unreadable" };
      }
    },
  };
}

export interface HealthRouterOptions {
  readonly config: SylConfig;
  /** Extra dependencies to report. The billing check is always included. */
  readonly probes?: readonly HealthProbe[];
  readonly clock?: Clock;
  /**
   * What this process was built from, read once at boot from the stamp inside
   * `dist/`. `null` — the default — is what running from source looks like, and
   * is a real answer rather than a missing one.
   *
   * Deliberately a value and not a function. It cannot change while the process
   * lives, and anything that recomputed it per request would be asking the
   * working tree rather than the artifact, which is the exact mistake this
   * field exists to make impossible.
   */
  readonly build?: BuildInfo | null;
  /**
   * How many conversations have a turn running or queued, asked at request
   * time. Omit for a service with no conversation surface.
   */
  readonly turnsInFlight?: () => number;
}

/**
 * Build the health router.
 *
 * `startedAt` is captured when the router is built, which is once, during
 * boot — so it is the service's start time rather than the first request's.
 */
export function createHealthRouter(options: HealthRouterOptions): Router {
  const { config } = options;
  const clock = options.clock ?? systemClock;
  const probes: readonly HealthProbe[] = [subscriptionRailsProbe(config), ...(options.probes ?? [])];
  const startedAt = instant(clock());
  const build = options.build ?? null;
  const turnsInFlight = options.turnsInFlight;

  const router = Router();

  router.get("/health", (_request, response) => {
    const checks: HealthCheck[] = probes.map((probe) => {
      const result = probe.run();
      return { name: probe.name, status: result.status, detail: result.detail ?? null };
    });

    const body: HealthStatus = {
      status: worstStatus(checks),
      version: config.version,
      startedAt,
      now: instant(clock()),
      checks,
      // Always present, explicitly null when unknown. A field that disappears
      // when there is nothing to say is indistinguishable from an old build
      // that never had the field at all — which is the one question this
      // endpoint must be able to answer.
      build,
      ...(turnsInFlight === undefined ? {} : { turnsInFlight: turnsInFlight() }),
    };

    // Always 200. `status` inside the body is the answer; a non-200 here would
    // be indistinguishable from the proxy in front of Syl being unhappy, and
    // "degraded" is information a monitor should read, not a failure it should
    // retry through.
    sendOk(response, body);
  });

  return router;
}
